src/coreclr/jit/importercalls.cpp

```diff
@@ -120,25 +123,192 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
+
+        if (instance->IsIconHandle(GTF_ICON_OBJ_HDL))
+        {
+            // Frozen delegates should never point to collectible contexts.
+            object = (void*)instance->AsIntCon()->gtIconVal;
+            JITDUMP("impImportCall trying to transform delegate - found frozen\n");
+        }
+        else if (instance->OperIs(GT_IND))
+        {
+            GenTree* address = instance->AsIndir()->gtOp1;
+            if (address->IsIconHandle(GTF_ICON_CONST_PTR))
+            {
+                // Const static addresses only exist in non-unloadable contexts,
+                // such context will keep the delegate permanently alive so even
+                // if it points to an unloadable method, we can still handle it safely.
+                // This lets us inline DynamicMethods.
+                indirect = true;
+                object = (void*)address->AsIntCon()->gtIconVal;
+                JITDUMP("impImportCall trying to transform delegate - found static readonly\n");
+            }
+        }
+        // TODO-Perf: Handle unloadable delegates in the same context.
+
+        if (object != nullptr)
+        {
+            JITDUMP("impImportCall trying to transform delegate - found delegate at 0x%llx, indirect: %d\n", object,
+                    indirect ? 1 : 0);
+            replacementMethod = info.compCompHnd->getMethodFromDelegate(object, indirect);
+
+            if (replacementMethod != NO_METHOD_HANDLE)
+            {
+                firstArgPos = (int)(stackState.esStackDepth - originalSig.numArgs);
+                assert(firstArgPos >= 0);
+                newThis =
+                    gtNewIndir(TYP_REF,
+                               gtNewOperNode(GT_ADD, TYP_BYREF, instance,
+                                             gtNewIconNode(eeGetEEInfo()->offsetOfDelegateFirstTarget, TYP_I_IMPL)),
+                               GTF_IND_INVARIANT | GTF_IND_NONNULL | GTF_IND_NONFAULTING);
+            }
+        }
+    }
+
+    if (replacementMethod != NO_METHOD_HANDLE)
+    {
+        JITDUMP("impImportCall trying to transform call - found target method %s\n",
+                eeGetMethodName(replacementMethod));
+        CORINFO_SIG_INFO     methodSig;
+        CORINFO_CLASS_HANDLE targetClass = info.compCompHnd->getMethodClass(replacementMethod);
+        info.compCompHnd->getMethodSig(replacementMethod, &methodSig, targetClass);
+
+        bool     removeInstance   = false;
+        unsigned replacementFlags = info.compCompHnd->getMethodAttribs(replacementMethod);
+
+        if ((replacementFlags & CORINFO_FLG_PINVOKE) != 0)
+        {
+            JITDUMP("impImportCall aborting transformation - found PInvoke\n");
+        }
+        else if (methodSig.hasImplicitThis() && eeIsValueClass(targetClass))
+        {
+            JITDUMP("impImportCall aborting transformation - valuetype instance methods are not supported\n");
+        }
+        else if (!impCanSubstituteSig(&originalSig, &methodSig, &removeInstance))
```

Comment:
Would it be better if the JIT/EE interface method returns the information about the transformation that needs to happen as part of the delegate call?

It is hard to tell whether this logic is always going to produce the same result as the VM logic that does the same.

Reply:
> the VM logic that does the same

What VM logic are you referring to? Do you mean `IsMethodDescCompatible`? If so, do I understand it correctly that you're suggesting to use that for signature checking for delegates and only use the JIT signature check for `calli`?

I actually had some simple logic for this in the VM for this but I've moved it back to the JIT before pushing cause I didn't see any simple way to do it more properly in the VM.

Reply:
I would expect this logic to be similar to the logic used to construct delegates: https://github.com/dotnet/runtime/blob/f7334fab4babc2cef606fafc053b4e5abe4edf80/src/coreclr/vm/comdelegate.cpp#L1731-L1801

Your current code seems to be doing more than what's required - it should not be necessary to compare all arguments, delegate constructor does not do that either. It may be also missing handing of the less common cases like delegates that point to VSD stubs, but I am not able to tell for sure.

Reply:
> Your current code seems to be doing more than what's required - it should not be necessary to compare all arguments

We need to verify the method matches the delegate to avoid crashing while JIT-ting when somebody does an invalid `Unsafe.As` in an unreachable path.

> delegate constructor does not do that either

From what I can see, reflection delegate creation does that via `IsMethodDescCompatible`.

Reply:
> We need to verify the method matches the delegate to avoid crashing while JIT-ting when somebody does an invalid Unsafe.As in an unreachable path.

The existing virtual method devitalization handles this case by just checking whether the target is castable to the desired type (see [CORINFO_DEVIRTUALIZATION_FAILED_CAST](https://github.com/search?q=repo%3Adotnet%2Fruntime%20CORINFO_DEVIRTUALIZATION_FAILED_CAST&type=code) ). You should be able to do the same here.

Reply:
> From what I can see, reflection delegate creation does that via IsMethodDescCompatible.

Reflection API needs to validate the arguments. JIT does need to validate the IL. It is fine for JIT to assume valid IL and crash otherwise.

Reply:
> > From what I can see, reflection delegate creation does that via IsMethodDescCompatible.
> 
> Reflection API needs to validate the arguments. JIT does need to validate the IL. It is fine for JIT to assume valid IL and crash otherwise.

We need to handle dynamically invalid code if it's never reachable though, this can't crash if the environment variable is or isn't defined:
```C#
public const string Variable = "EnvVar";
public static readonly Func<int> A = () => 5;
public static readonly Func<int, int> B = (i) => i + 5;
public static readonly Delegate D = Environment.GetEnvironmentVariable(Variable) ? A : B;

public static int Method()
{
    if (Environment.GetEnvironmentVariable(Variable))
        return Unsafe.As<Func<int>>(D)();
    return Unsafe.As<Func<int, int>>(D)(5);
}
```

Reply:
Right, I mean that you can do this by just calling CanCast instead of comparing the signature.

For example, for `Unsafe.As<Func<int>>(D)();`: you take the actual type of D and the delegate type (`Func<int>` in this case) and verify whether the actual type of D can be cast to the delegate type.

Reply:
Oh wait, you mean target as in the delegate, I thought you were referring to the `_target` field with the instance.

Reply:
Done but is it just me or [this check](https://github.com/dotnet/runtime/blob/1e3544ec04e27538f3d38d274f7a65277f0d681f/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DelegateCreationInfo.cs#L210) in NativeAOT is wrong? AFAIR it's open when it says it's closed.